test/unit/uk/gov/hmrc/tariffclassificationfrontend/controllers/ReferCaseControllerSpec.scala

```diff
@@ -0,0 +1,188 @@
+
+      status(result) shouldBe Status.OK
+      contentTypeOf(result) shouldBe Some(MimeTypes.HTML)
+      charsetOf(result) shouldBe Some("utf-8")
+      bodyOf(result) should include("This case has been referred")
+    }
+
+    "redirect to Application Details for non OPEN statuses" in {
+      when(casesService.getOne(refEq("reference"))(any[HeaderCarrier])).thenReturn(successful(Some(caseWithStatusNEW)))
+
+      val result: Result= await(controller.confirmReferCase("reference")(newFakePOSTRequestWithCSRF("queue")))
+
+      status(result) shouldBe Status.SEE_OTHER
+      contentTypeOf(result) shouldBe None
+      charsetOf(result) shouldBe None
+      locationOf(result) shouldBe Some("/tariff-classification/cases/reference/application")
+    }
+
+    "return Not Found and HTML content type on missing Case" in {
+      when(casesService.getOne(refEq("reference"))(any[HeaderCarrier])).thenReturn(successful(None))
+      val result: Result = await(controller.confirmReferCase("reference")(newFakePOSTRequestWithCSRF("queue")))
+
+      status(result) shouldBe Status.OK
+      contentTypeOf(result) shouldBe Some(MimeTypes.HTML)
+      charsetOf(result) shouldBe Some("utf-8")
+      bodyOf(result) should include("We could not find a Case with reference")
+    }
+
+    "propagate the error in case the CaseService fails to release the case" in {
+      val error = new IllegalStateException("expected error")
+
+      when(casesService.getOne(refEq("reference"))(any[HeaderCarrier])).thenReturn(failed(error))
+
+      val caught = intercept[error.type] {
+        await(controller.confirmReferCase("reference")(newFakePOSTRequestWithCSRF("queue")))
+      }
+      caught shouldBe error
+    }
+  }
+
+  private def newFakeGETRequestWithCSRF(): FakeRequest[AnyContentAsEmpty.type] = {
+    val tokenProvider: TokenProvider = app.injector.instanceOf[TokenProvider]
+    val csrfTags = Map(Token.NameRequestTag -> "csrfToken", Token.RequestTag -> tokenProvider.generateToken)
+    FakeRequest("GET", "/", FakeHeaders(), AnyContentAsEmpty, tags = csrfTags)
+  }
+
+  private def newFakePOSTRequestWithCSRF(queue: String): FakeRequest[AnyContentAsFormUrlEncoded] = {
+    val tokenProvider: TokenProvider = app.injector.instanceOf[TokenProvider]
+    val csrfTags = Map(Token.NameRequestTag -> "csrfToken", Token.RequestTag -> tokenProvider.generateToken)
+    FakeRequest("POST", "/", FakeHeaders(), AnyContentAsFormUrlEncoded, tags = csrfTags).withFormUrlEncodedBody("queue" -> queue)
+  }
+
+  private def newInvalidFakePOSTRequestWithCSRF(): FakeRequest[AnyContentAsFormUrlEncoded] = {
+    val tokenProvider: TokenProvider = app.injector.instanceOf[TokenProvider]
+    val csrfTags = Map(Token.NameRequestTag -> "csrfToken", Token.RequestTag -> tokenProvider.generateToken)
+    FakeRequest("POST", "/", FakeHeaders(), AnyContentAsFormUrlEncoded, tags = csrfTags).withFormUrlEncodedBody()
+  }
+
+  private def locationOf(result: Result): Option[String] = {
```

Comment:
I believe most of these private methods are used also by the other controller test classes.
You could create a trait and put those methods in there, with `protected` access modifier.
Then all the controller test classes can just extend (`with`) this new trait you created.